packages/flutter/lib/src/painting/borders.dart

```diff
@@ -259,6 +259,318 @@ class BorderSide {
   String toString() => '$runtimeType($color, ${width.toStringAsFixed(1)}, $style)';
 }
 
+/// Base class for shape outlines.
+///
+/// This class handles how to add multiple borders together.
+@immutable
+abstract class ShapeBorder {
+  /// Abstract const constructor. This constructor enables subclasses to provide
+  /// const constructors so that they can be used in const expressions.
+  const ShapeBorder();
+
+  /// The widths of the sides of this border represented as an [EdgeInsets].
+  ///
+  /// Specifically, this is the amount by which a rectangle should be inset so
+  /// as to avoid painting over any important part of the border. It is the
+  /// amount by which additional borders will be inset before they are drawn.
+  ///
+  /// This can be used, for example, with a [Padding] widget to inset a box by
+  /// the size of these borders.
+  ///
+  /// Shapes that have a fixed ratio regardless of the area on which they are
+  /// painted, or that change their rendering based on the size, for instance
+  /// [CircleBorder], will not return invalid [dimensions] information (because
+  /// they cannot know their eventual size when computing their dimensions).
+  EdgeInsetsGeometry get dimensions;
+
+  /// Attempts to create a new object that represents the amalgamation of [this]
+  /// border and the `other` border.
+  ///
+  /// If the type of the other border isn't known, or the given instance cannot
+  /// be reasonably added to this instance, then this should return null.
+  ///
+  /// This method is used by the [operator +] implementation.
+  ///
+  /// The `reversed` argument is true if this object was the right operand of
+  /// the `+` operator, and false if it was the left operand.
+  @protected
+  ShapeBorder add(ShapeBorder other, { bool reversed: false }) => null;
+
+  /// Creates a new border consisting of the two borders on either side of the
+  /// operator.
+  ///
+  /// If the borders belong to classes that know how to add themselves, then
+  /// this results in a new border that represents the intelligent addition of
+  /// those two borders (see [add]). Otherwise, an object is returned that
+  /// merely paints the two borders sequentially, with the left hand operand on
+  /// the inside and the right hand operand on the outside.
+  ShapeBorder operator +(ShapeBorder other) {
+    return add(other) ?? other.add(this, reversed: true) ?? new _CompoundBorder(<ShapeBorder>[other, this]);
+  }
+
+  /// Creates a new border with the widths of this border multiplied by `t`.
```

Comment:
Any (finite) value is valid. I'm reluctant to add this information here because if we add it here we should add it to all other scale (and lerp?) methods, but maybe we should do that (in another PR)? Please file a bug if you think we should do that, and I'll make sure it's done.